security-baseline/rule-types/github/osps-qa-03.yaml

```
@@ -0,0 +1,73 @@
+---
+version: v1
+release_phase: alpha
+type: rule-type
+name: osps-qa-03
+display_name: Package management file listing dependencies is present
+short_failure_message: No package management file listing dependencies was found
+severity:
+  value: info
+context:
+  provider: github
+description: |
+  This rule ensures that the repository provides a dependency list that accounts 
+  for the direct language dependencies when the package management system supports it.
+  It checks for the presence of a Gemfile.lock, go.sum, package-lock.json or 
+  Cargo.toml according to the package management system used.
+guidance: |
+  Ensure that the repository provides a dependency list that accounts for the direct
+  language dependencies in the form of a lockfile, for example Gemfile.lock, go.sum,
+  package-lock.json etc.
+  Ensure the lockfile is in the same directory as the package file.
+def:
+  in_entity: repository
+  rule_schema: {}
+  ingest:
+    type: git
+  eval:
+    type: rego
+    rego:
+      type: deny-by-default
+      def: |
+        package minder
+
+        import rego.v1
+
+        default allow := false
+        default skip := false
+        default message := "Cannot find lockfile in the same directory as the package file"
+        
+        package_manager_files := [
+          {"name": "Gemfile", "lockfiles": ["Gemfile.lock"]},
+          {"name": "go.mod", "lockfiles": ["go.sum"]},
+          {"name": "package.json", "lockfiles": ["package-lock.json", "yarn.lock"]},
+          {"name": "Cargo.toml", "lockfiles": ["Cargo.lock"]},
+        ]
+        
+        skip if {
+          # Skip if no package manager file exists
+          every package_manager in package_manager_files {
+            required_files := file.ls_glob(sprintf("./%s", [package_manager.name]))
```

Unfortunately the glob library we're using in Minder doesn't support `**` patterns, which is why this only looks at the root directory. This is probably something we want to change in the Minder code.

---

Yeah -- we could also use `file.walk` for this:

```suggestion
            all_files := file.walk(".")
            package_manager_files = [ filepath |
              filepath := all_files[_];
              glob.match("**/" + package_manager[_].name, ["/"], filepath)
            ]
```